data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,173 @@
+import software.amazon.kinesis.lifecycle.events.ProcessRecordsInput;
+import software.amazon.kinesis.lifecycle.events.ShardEndedInput;
+import software.amazon.kinesis.lifecycle.events.ShutdownRequestedInput;
+import software.amazon.kinesis.processor.RecordProcessorCheckpointer;
+import software.amazon.kinesis.processor.ShardRecordProcessor;
+import software.amazon.kinesis.retrieval.KinesisClientRecord;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+public class KinesisRecordProcessor implements ShardRecordProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisSource.class);
+    private final StreamIdentifier streamIdentifier;
+    private final KinesisStreamConfig kinesisStreamConfig;
+    private final Duration checkpointInterval;
+    private final KinesisSourceConfig kinesisSourceConfig;
+    private final Buffer<Record<Event>> buffer;
+    private String kinesisShardId;
+    private final InputCodec codec;
+    private long lastCheckpointTimeInMillis;
+    private final int bufferTimeoutMillis;
+    private final AcknowledgementSetManager acknowledgementSetManager;
+    private final Counter acknowledgementSetCallbackCounter;
+    private final Counter recordProcessingErrors;
+    private final Counter checkpointFailures;
+    private static final Duration ACKNOWLEDGEMENT_SET_TIMEOUT = Duration.ofSeconds(20);
+    public static final String ACKNOWLEDGEMENT_SET_CALLBACK_METRIC_NAME = "acknowledgementSetCallbackCounter";
+    public static final String KINESIS_RECORD_PROCESSING_ERRORS = "recordProcessingErrors";
+    public static final String KINESIS_CHECKPOINT_FAILURES = "checkpointFailures";
+    public static final String KINESIS_STREAM_TAG_KEY = "stream";
+
+    public KinesisRecordProcessor(Buffer<Record<Event>> buffer,
+                                  final KinesisSourceConfig kinesisSourceConfig,
+                                  final AcknowledgementSetManager acknowledgementSetManager,
+                                  final PluginMetrics pluginMetrics,
+                                  final PluginFactory pluginFactory,
+                                  final StreamIdentifier streamIdentifier) {
+        this.bufferTimeoutMillis = (int) kinesisSourceConfig.getBufferTimeout().toMillis();
+        this.streamIdentifier = streamIdentifier;
+        this.kinesisSourceConfig = kinesisSourceConfig;
+        this.kinesisStreamConfig = getStreamConfig(kinesisSourceConfig);
+        final PluginModel codecConfiguration = kinesisSourceConfig.getCodec();
+        final PluginSetting codecPluginSettings = new PluginSetting(codecConfiguration.getPluginName(), codecConfiguration.getPluginSettings());
+        this.codec = pluginFactory.loadPlugin(InputCodec.class, codecPluginSettings);
```

Comment:
Might KCL create multiple of these processors? We can create the `InputCodec` once, either in `KinesisService` (probably the ideal location) or `KinesisSharedRecordProcessorFactory`. Then pass it in here. This way, multiple instances of `KinesisRecordProcessor` all share one instance of `InputCodec`.

This also prepares us to use #4838 once implemented.

Reply:
Thanks @dlvenable. I will pass the `InputCodec` created once in `KinesisService` to the `KinesisSharedRecordProcessorFactory`. This way we can have one single instance of `InputCodec` passed to all processors.